[GSC] Add base-image's environment variables as part of gsc entrypoint.mani…

veenasai2 opened the pull request:
**Description of the changes**

Previously gsc entrypoint.manifest file did not contain docker base image specific environment variables. Only "PATH" environment variable is fetched as part of finalize_manifest.py and integrated into entrypoint.manifest file.

We can see base image's environment variables using:
docker inspect

This PR extracts environment variables from base image and integrates them into entrypoint.manifest file.

Sample environment variables added in the entrypoint.manifest
loader.env.PYTHON_VERSION="3.8.7"
loader.env.PYTHON_PIP_VERSION="21.0"

**How to test this PR?**
(1) GSC in Jenkin must succeed.
(2) During gsc build we can print the contents of entrypoint.manifest from finalize_manifest.py main() function and verify that all environment variables are successfully added.

**Note:** This PR is linked to "https://github.com/oscarlab/graphene/pull/2174". I have addressed the review comments here.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/2195)
<!-- Reviewable:end -->

dimakuv reviewed (approved):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel)



<!-- Sent from Reviewable.io -->

dimakuv commented:
Jenkins, test this please

dimakuv reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @veenasai2)

___
*[Tools/gsc/gsc.py, line 176 at r1](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUh3qkFEIc2ZCXiIp8j:-MUh3qkFEIc2ZCXiIp8k:b-kk2ebt) ([raw file](https://github.com/oscarlab/graphene/blob/33553478f13c60df227f68ce524c31ac59cb12a4/Tools/gsc/gsc.py#L176)):*
> ```Python
> 
>     # extract base docker image's environment variables to append inside entrypoint.manifest file
>     base_image_environment=extract_environment_from_image_config(original_image.attrs['Config'])
> ```

Our syntax checker complained about this line:
```
C:176, 0: Exactly one space required around assignment
    base_image_environment=extract_environment_from_image_config(original_image.attrs['Config'])
                          ^ (bad-whitespace)
```

Please add spaces around the `=` sign.


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 176 at r1](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUh3qkFEIc2ZCXiIp8j:-MUh4v4f216mRvRmbLJv:bfuvxil) ([raw file](https://github.com/oscarlab/graphene/blob/33553478f13c60df227f68ce524c31ac59cb12a4/Tools/gsc/gsc.py#L176)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Our syntax checker complained about this line:
```
C:176, 0: Exactly one space required around assignment
    base_image_environment=extract_environment_from_image_config(original_image.attrs['Config'])
                          ^ (bad-whitespace)
```

Please add spaces around the `=` sign.
</blockquote></details>

ok sure. Only this change required or should I wait for Jenkins results as well?


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @veenasai2)

___
*[Tools/gsc/gsc.py, line 176 at r1](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUh3qkFEIc2ZCXiIp8j:-MUh7AP54QV15MafUrTB:bfkykkw) ([raw file](https://github.com/oscarlab/graphene/blob/33553478f13c60df227f68ce524c31ac59cb12a4/Tools/gsc/gsc.py#L176)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

ok sure. Only this change required or should I wait for Jenkins results as well?
</blockquote></details>

Looks like only this change. I'll re-trigger Jenkins anyway, after you push a fix.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUmHaZCBlYnvJicFARD:-MUmHaZCBlYnvJicFARE:b-9vlagy) (no related file):*
@veenasai2 In the future, after you git-push your changes, please use Reviewable to mark the comments as `Done` (or write your comments/questions if not yet done).

---


<!-- Sent from Reviewable.io -->

dimakuv commented:
Jenkins, test this please

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 176 at r1](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUh3qkFEIc2ZCXiIp8j:-MUmIiOABO6t3x3gPhPm:b-7rl0jd) ([raw file](https://github.com/oscarlab/graphene/blob/33553478f13c60df227f68ce524c31ac59cb12a4/Tools/gsc/gsc.py#L176)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Looks like only this change. I'll re-trigger Jenkins anyway, after you push a fix.
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

veenasai2 commented:
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUmHaZCBlYnvJicFARD:-MUmIoF2AKyWhANfxF93:bijd0mr) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@veenasai2 In the future, after you git-push your changes, please use Reviewable to mark the comments as `Done` (or write your comments/questions if not yet done).
</blockquote></details>

sure, thank you.

---


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @veenasai2)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriRPT3RNpuJfplopp:-MUriRPT3RNpuJfplopq:b65rcii) (no related file):*
Please add [GSC] component tag to the commit message (we can do this during the final rebase)

---
___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MUriwwO1Ehr5jDYeGvd:b-sgb4fr) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
> ```Python
>         # the need for multi-line envvars arises
>         if '\n' in env_var:
>             # we use TOML's basic single-line strings, can't have newlines
> ```

Please print a warning here that an env var was skipped (with proper explanation for the user)
___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MUrjDgh0kfxuYruiQts:b-woumci) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
> ```
> '\\': r'\\'
> ```

Nitpick:  `r'\': r'\\'` would be easier to read IMO
___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MUrkfo457bhF8rE3cmS:b1sw8l2) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
> ```Python
>         env_var_name = escaped_env_var.split('=')[0]
>         if env_var_name == f'PATH' or env_var_name == f'LD_LIBRARY_PATH':
>             # PATH and LD_LIBRARY_PATH are already part of entrypoint.manifest.template.
> ```

I don't like hardcoding here which variables are set and which are not in the template. What if some other template uses different variables?
I'd rather change `finalize_manifest.py` to append to the variables it uses if they already exist and for others just error out that there's a conflict between the template and what user defined in the configuration.
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MUrkWrJBn9BIrlHysOK:bvm012b) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
> ```Python
>             # Their values are provided in finalize_manifest.py, hence skipping here.
>             continue
>         env_var_value = escaped_env_var.split('=')[1]
> ```

This won't work if the value contains `=`. You can fix it using `maxsplit=1`.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow and @veenasai2)

___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MUrmTxu03EbX1riG7s7:bcwfzca) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please print a warning here that an env var was skipped (with proper explanation for the user)
</blockquote></details>

+1, something like:
```
print(f'Skipping environment variable `{env_var.split('=')[0]}`: its value contains newlines.')
```
___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MUrn8pp2lu5qTFaCJeD:b-vkxaxc) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> ```
> '\\': r'\\'
> ```

Nitpick:  `r'\': r'\\'` would be easier to read IMO
</blockquote></details>

+1. This was copied from `finalize_manifest.py`, so please apply the same in that file.
___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MUrnIN4Av0yAF0Pua2h:b-i0d55o) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
@mkow I disagree. This will require adding a proper TOML parser to all our Python scripts (including the normal Graphene ones to create and sign the enclave). We currently don't have such a parser, and this is out of scope of this PR (actually, this will be a really big chunk of work).

> the variables it uses if they already exist

How do you find the variables that already exist in the manifest? We currently do heuristics everywhere (like `string.startswith()`), but this is an artifact of the past, when we didn't have TOML. With TOML, a variable like `loader.env.PATH` can actually be declared in several TOML ways, so without the proper TOML parser we're doomed. Imagine `[loader] [env] PATH = """blabla"""`, how are we supposed to check this (valid) TOML entry?
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MUro8upDz6LVTLgUHki:b32v57) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

This won't work if the value contains `=`. You can fix it using `maxsplit=1`.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MUt2OL69kYWR-wrP7pW:b77i4af) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

+1, something like:
```
print(f'Skipping environment variable `{env_var.split('=')[0]}`: its value contains newlines.')
```
</blockquote></details>

sure, thanks.


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MUt4zWn3gKFSTholQt5:b822ugu) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@mkow I disagree. This will require adding a proper TOML parser to all our Python scripts (including the normal Graphene ones to create and sign the enclave). We currently don't have such a parser, and this is out of scope of this PR (actually, this will be a really big chunk of work).

> the variables it uses if they already exist

How do you find the variables that already exist in the manifest? We currently do heuristics everywhere (like `string.startswith()`), but this is an artifact of the past, when we didn't have TOML. With TOML, a variable like `loader.env.PATH` can actually be declared in several TOML ways, so without the proper TOML parser we're doomed. Imagine `[loader] [env] PATH = """blabla"""`, how are we supposed to check this (valid) TOML entry?
</blockquote></details>

Based on this discussion, can we take up this specific requirement for future PRs?


<!-- Sent from Reviewable.io -->

veenasai2 commented:
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriRPT3RNpuJfplopp:-MUt5wc28CRGX94We9-d:b77i4af) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please add [GSC] component tag to the commit message (we can do this during the final rebase)
</blockquote></details>

sure, thanks.

---


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MUtHeYd92Iv9fAAUPwC:b-x9y34f) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

+1. This was copied from `finalize_manifest.py`, so please apply the same in that file.
</blockquote></details>

based on my reading:  r'\' is invalid raw string. Here the end quote is missing, because it will be escaped by backslash.  


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MUtJyKnD_PI3D3Djphh:b-byjv9t) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

based on my reading:  r'\' is invalid raw string. Here the end quote is missing, because it will be escaped by backslash.  
</blockquote></details>

I mean r' \ ' is invalid raw string.  


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MUwwR0nChmgMAsYdsFo:b-nx718o) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
I think @veenasai2 is right, the way suggested by Michal is not working (unless I'm doing something wrong in my simply Python test):
```python
$ python3

# the current way works
>>> print(str("hello\world".translate(str.maketrans({'\\': r'\\', '"': r'\"'}))))
hello\\world

# the proposed by Michal doesn't work
>>> print(str("hello\world".translate(str.maketrans({r'\': r'\\', '"': r'\"'}))))
  File "<stdin>", line 1
    print(str("hello\world".translate(str.maketrans({r'\': r'\\', '"': r'\"'}))))
                                                                                ^
SyntaxError: unexpected character after line continuation character
```

@mkow What do you think?


<!-- Sent from Reviewable.io -->

mkow reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv, @mkow, and @veenasai2)

___
*[Tools/gsc/gsc.py, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrjDgh0kfxuYruiQtr:-MVGfVmJ6cc_5A2fz34l:b-xon1mx) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L99)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I think @veenasai2 is right, the way suggested by Michal is not working (unless I'm doing something wrong in my simply Python test):
```python
$ python3

# the current way works
>>> print(str("hello\world".translate(str.maketrans({'\\': r'\\', '"': r'\"'}))))
hello\\world

# the proposed by Michal doesn't work
>>> print(str("hello\world".translate(str.maketrans({r'\': r'\\', '"': r'\"'}))))
  File "<stdin>", line 1
    print(str("hello\world".translate(str.maketrans({r'\': r'\\', '"': r'\"'}))))
                                                                                ^
SyntaxError: unexpected character after line continuation character
```

@mkow What do you think?
</blockquote></details>

Wow, I'm actually surprised here, because backslashes are treated literally in raw strings, and e.g. `r'\\'` produces a string with two backslashes. I guess this is a quirk in Python's parsing grammar.
Anyways, I'm unblocking this.
___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MVGgNNTCXEs7E-bHPfG:b-9ldchw) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
> We currently don't have such a parser

This got outdated pretty fast :) See #2210. But except this it's a good point, in current implementation it's impossible.

So, do we want to wait for #2210 and implement it properly, or go with current, hacky solution? #2210 is blocking two other issues, so it should be merged soon.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MVGotPf21ZookwcXyCr:b3trwih) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> We currently don't have such a parser

This got outdated pretty fast :) See #2210. But except this it's a good point, in current implementation it's impossible.

So, do we want to wait for #2210 and implement it properly, or go with current, hacky solution? #2210 is blocking two other issues, so it should be merged soon.
</blockquote></details>

I would suggest to unblock this PR. There is a need to update all GSC scripts to a TOML parser, not only this one. So this seems like a task for a separate PR.

Also, are we sure that #2210 solves all TOML-parser issues in Python? I thought we have a couple other Python scripts that need to be updated...


<!-- Sent from Reviewable.io -->

mkow reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

___
*[Tools/gsc/gsc.py, line 102 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkfo457bhF8rE3cmR:-MVGsOSq2A1WRNQpWkdD:b-nuwkrc) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L102)):*
Ok, let's unblock it then.

> I thought we have a couple other Python scripts that need to be updated...

In the core Graphene only sgx_sign.py parsed TOML. I'm not sure about GSC though.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @veenasai2)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MVH45Rj2g1Mo56w92Ri:-MVH45Rj2g1Mo56w92Rj:b-3p24n6) (no related file):*
@veenasai2 Just a reminder to fix the remaining two issues in this PR (after that we can rebase it and merge).

---


<!-- Sent from Reviewable.io -->

veenasai2 commented:
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2195#-MVH45Rj2g1Mo56w92Ri:-MVKRWoX4UyF_2jqpHKX:bb7f2mf) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@veenasai2 Just a reminder to fix the remaining two issues in this PR (after that we can rebase it and merge).
</blockquote></details>

done

---


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MUtCa-oATFQnpkegy12:b-7rl0kd) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

sure, thanks.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVKRcPe36B2IvJWa_Ok:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

 sure, will do that soon. Thanks.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MVLlXsk7Vab0hsabebx:bq5gbgn) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

done
</blockquote></details>

Could you wrap it? We have 100 chars per line limit.
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVLle2g2YbXRd4k5K5-:b-bcxgvz) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Could you use named arg for the second argument? I.e. `maxsplit=1` (both here and above)


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVLmhEl2xg_G06s6Je4:b-m2nyf6) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Could you use named arg for the second argument? I.e. `maxsplit=1` (both here and above)
</blockquote></details>

https://github.com/oscarlab/graphene/blob/master/Tools/gsc/gsc.py#L95   . Should I make a change here as well? 


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MVLnSeF3wWWgzDpZ-9t:bkoeezw) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Could you wrap it? We have 100 chars per line limit.
</blockquote></details>

Currently the print message is 96 char long. Will it be ok, if I insert a newline or another print statement? 
or 
print(f'Skipping env var `{env_var.split("=")[0]}`: its value contains newlines.')


<!-- Sent from Reviewable.io -->

mkow reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MVLqxeW9lSPDoypOmFH:buptcw5) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

Currently the print message is 96 char long. Will it be ok, if I insert a newline or another print statement? 
or 
print(f'Skipping env var `{env_var.split("=")[0]}`: its value contains newlines.')
</blockquote></details>

I meant the source code, not the message itself:
```Python
            print(f'Skipping environment variable `{env_var.split("=", 1)[0]}`: its value contains '
                  'newlines.')
```
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVLrkwl5A3fNB0zNrFm:bczfn3f) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

https://github.com/oscarlab/graphene/blob/master/Tools/gsc/gsc.py#L95   . Should I make a change here as well? 
</blockquote></details>

Not necessarily (as the other place is not related to the PR), but if you want you can fix both.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: 0 of 1 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @veenasai2)

___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVM0ogr3JkjgNQ8KWyZ:bppe4q1) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Not necessarily (as the other place is not related to the PR), but if you want you can fix both.
</blockquote></details>

Yes, it would be nice if you also fix that other place as well.

Just to make it clear, what Michal wants is this:
```
... env_var.split("=", maxsplit=1)[0] ...
env_var_name = escaped_env_var.split('=', maxsplit=1)[0]
env_var_value = escaped_env_var.split('=', maxsplit=1)[1]
```

(So, in all cases where you do a split on `"="`, please add `maxsplit=` before the second argument)


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 97 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUriwwO1Ehr5jDYeGvc:-MVLteRPET_94Ab27MGn:b-zh5tqo) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L97)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I meant the source code, not the message itself:
```Python
            print(f'Skipping environment variable `{env_var.split("=", 1)[0]}`: its value contains '
                  'newlines.')
```
</blockquote></details>

ok, thank you. Done now.


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVM1Nzk-6HZa03cqDKN:b-uo4wbe) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Yes, it would be nice if you also fix that other place as well.

Just to make it clear, what Michal wants is this:
```
... env_var.split("=", maxsplit=1)[0] ...
env_var_name = escaped_env_var.split('=', maxsplit=1)[0]
env_var_value = escaped_env_var.split('=', maxsplit=1)[1]
```

(So, in all cases where you do a split on `"="`, please add `maxsplit=` before the second argument)
</blockquote></details>

ok, thank you. 


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @veenasai2)

___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVM4SDGEX2Nhfsl0iU4:b-pwvbbm) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

ok, thank you. 
</blockquote></details>

Sorry, still slightly wrong: our Python standard doesn't allow `maxsplit = 1` (with spaces around). Our Python standard forces `maxsplit=1` (without spaces around). This standard is forcibly applied in our Jenkins for all function arguments in our Python scripts.

This is some weird Python convention, to be honest, I don't remember why this is important. But we're just following it.

(You may wonder -- in our C code, we have the opposite standard. Why is this so? I don't know, Python coding style is weird to me. So we end up having slightly different conventions for our C code and for our Python code. Probably others know more why this is so.)
___
*[Tools/gsc/gsc.py, line 98 at r4](https://reviewable.io/reviews/oscarlab/graphene/2195#-MVM4JdIA1kySbxpjQNg:-MVM4JdIA1kySbxpjQNh:bh4r56q) ([raw file](https://github.com/oscarlab/graphene/blob/37d62b22d4d08f31741576536b89c4377ef4daad/Tools/gsc/gsc.py#L98)):*
> ```Python
>         if '\n' in env_var:
>             # we use TOML's basic single-line strings, can't have newlines
>             print(f'Skipping environment variable `{env_var.split("=", maxsplit = 1)[0]}`: ' 
> ```

There is a redundant space at the end of the line :(


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @veenasai2)

___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVM5Ogz5XfPbBgJQGvZ:b-1mndb4) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Sorry, still slightly wrong: our Python standard doesn't allow `maxsplit = 1` (with spaces around). Our Python standard forces `maxsplit=1` (without spaces around). This standard is forcibly applied in our Jenkins for all function arguments in our Python scripts.

This is some weird Python convention, to be honest, I don't remember why this is important. But we're just following it.

(You may wonder -- in our C code, we have the opposite standard. Why is this so? I don't know, Python coding style is weird to me. So we end up having slightly different conventions for our C code and for our Python code. Probably others know more why this is so.)
</blockquote></details>

https://stackoverflow.com/questions/8853063/pep-8-why-no-spaces-around-in-keyword-argument-or-a-default-parameter-value -- I googled this weird Python convention. Well, it seems that the short answer is `Because Python author wanted so, deal with it`.


<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 98 at r4](https://reviewable.io/reviews/oscarlab/graphene/2195#-MVM4JdIA1kySbxpjQNg:-MVM669H1AbaK54KavJr:b-wpb374) ([raw file](https://github.com/oscarlab/graphene/blob/37d62b22d4d08f31741576536b89c4377ef4daad/Tools/gsc/gsc.py#L98)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

There is a redundant space at the end of the line :(
</blockquote></details>

ok, thanks.


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVM6TCa4wz6uuog4uUt:bvjajsb) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
> This is some weird Python convention, to be honest, I don't remember why this is important. But we're just following it.

I wouldn't say weird, it's pretty standard to have named args listed without spaces I think (I mean, in other languages also). Also, it makes it look distinct from assignments.


<!-- Sent from Reviewable.io -->

mkow reviewed (approved):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)

___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVMEy6o8W6-cIKk8bBb:bh03d1y) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, veenasai2 wrote…</i></summary><blockquote>

I just build gsc again and did docker run, did not face any issues because of spaces around maxsplit.  Also, printed the env variables from entrypoint.manifest , all seems fine (sample output:  loader.env.LANG = "C.UTF-8").  But , yeah if jenkins fails while parsing, I will remove the spaces around maxsplit. Thanks.
</blockquote></details>

@veenasai2: It's not Python which has a problem with the spaces, but our stylechecker (pylint). We use it to enforce consistent style everywhere.


<!-- Sent from Reviewable.io -->

mkow commented:
Jenkins, retest this please

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2195)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

veenasai2 commented:
___
*[Tools/gsc/gsc.py, line 105 at r2](https://reviewable.io/reviews/oscarlab/graphene/2195#-MUrkWrJBn9BIrlHysOJ:-MVMGng39iyUDYRRkkSk:b3lcvlm) ([raw file](https://github.com/oscarlab/graphene/blob/c3873e1995949019aa063b2956fb5276da441cdd/Tools/gsc/gsc.py#L105)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

@veenasai2: It's not Python which has a problem with the spaces, but our stylechecker (pylint). We use it to enforce consistent style everywhere.
</blockquote></details>

ok, thank you.


<!-- Sent from Reviewable.io -->
